lightning/src/ln/functional_test_utils.rs

```diff
@@ -1287,77 +1289,102 @@ pub fn send_payment<'a, 'b, 'c>(origin: &Node<'a, 'b, 'c>, expected_route: &[&No
-		if !skip_last || idx != expected_route.len() - 1 {
-			assert_eq!(events.len(), 1);
-			match events[0] {
-				MessageSendEvent::UpdateHTLCs { ref node_id, updates: msgs::CommitmentUpdate { ref update_add_htlcs, ref update_fulfill_htlcs, ref update_fail_htlcs, ref update_fail_malformed_htlcs, ref update_fee, ref commitment_signed } } => {
+	macro_rules! msgs_from_ev {
+		($ev: expr) => {
+			match $ev {
+				&MessageSendEvent::UpdateHTLCs { ref node_id, updates: msgs::CommitmentUpdate { ref update_add_htlcs, ref update_fulfill_htlcs, ref update_fail_htlcs, ref update_fail_malformed_htlcs, ref update_fee, ref commitment_signed } } => {
 					assert!(update_add_htlcs.is_empty());
 					assert!(update_fulfill_htlcs.is_empty());
 					assert_eq!(update_fail_htlcs.len(), 1);
 					assert!(update_fail_malformed_htlcs.is_empty());
 					assert!(update_fee.is_none());
-					expected_next_node = node_id.clone();
-					next_msgs = Some((update_fail_htlcs[0].clone(), commitment_signed.clone()));
+					((update_fail_htlcs[0].clone(), commitment_signed.clone()), node_id.clone())
 				},
 				_ => panic!("Unexpected event"),
 			}
-		} else {
-			assert!(events.is_empty());
-		}
-		if !skip_last && idx == expected_route.len() - 1 {
-			assert_eq!(expected_next_node, origin_node.node.get_our_node_id());
 		}
+	}
 
-		prev_node = node;
+	let mut per_path_msgs: Vec<((msgs::UpdateFailHTLC, msgs::CommitmentSigned), PublicKey)> = Vec::with_capacity(expected_paths.len());
+	let events = expected_paths[0].last().unwrap().node.get_and_clear_pending_msg_events();
+	assert_eq!(events.len(), expected_paths.len());
+	for ev in events.iter() {
+		per_path_msgs.push(msgs_from_ev!(ev));
 	}
+	per_path_msgs.sort_unstable_by(|(_, node_id_a), (_, node_id_b)| node_id_a.cmp(node_id_b));
+	expected_paths.sort_unstable_by(|path_a, path_b| path_a[path_a.len() - 2].node.get_our_node_id().cmp(&path_b[path_b.len() - 2].node.get_our_node_id()));
 
-	if !skip_last {
-		update_fail_dance!(origin_node, expected_route.first().unwrap(), true);
+	for (i, (expected_route, (path_msgs, next_hop))) in expected_paths.iter().zip(per_path_msgs.drain(..)).enumerate() {
+		let mut next_msgs = Some(path_msgs);
+		let mut expected_next_node = next_hop;
+		let mut prev_node = expected_route.last().unwrap();
 
-		let events = origin_node.node.get_and_clear_pending_events();
-		assert_eq!(events.len(), 1);
-		match events[0] {
-			Event::PaymentFailed { payment_hash, rejected_by_dest, .. } => {
-				assert_eq!(payment_hash, our_payment_hash);
-				assert!(rejected_by_dest);
-			},
-			_ => panic!("Unexpected event"),
+		for (idx, node) in expected_route.iter().rev().enumerate().skip(1) {
+			assert_eq!(expected_next_node, node.node.get_our_node_id());
+			if next_msgs.is_some() {
+				node.node.handle_update_fail_htlc(&prev_node.node.get_our_node_id(), &next_msgs.as_ref().unwrap().0);
+				let last_node = !(skip_last && idx == expected_route.len() - 1);
```

Comment:
This is equivalent to `!skip_last || idx != expected_route.len() - 1`, which I find more readable, and which is pretty different from "last node". I'm not really sure what is supposed to be going on here, but note that all tests pass if I set last_node to `idx == expected_route.len() - 1 && !skip_last` or, even, `idx == expected_route.len() - 1` (which I think is what really should be here).

Reply:
> note that all tests pass if I set last_node to idx == expected_route.len() - 1 && !skip_last or, even, idx == expected_route.len() - 1

I get test failures with both of those^ :thinking: 

> This is equivalent to !skip_last || idx != expected_route.len() - 1, which I find more readable, and which is pretty different from "last node".

That is more readable, and yeah the name was wrong. Took the suggestion and renamed

Reply:
> I get test failures with both of those^ thinking

lol sorry was running out of the coffee shop and ran cargo test in the wrong directory :facepalm:.